### PR TITLE
Add possibility to plot 2D hist with x and y projections

### PR DESCRIPTION
**Pull request by cyrraz**
Dear @0ctagon,

This PR adds the possibility to plot 2D hist with x and y projections.

The script to produce the figure in the documentation does not work yet, probably because of the use of a `save_as`, otherwise the function works. It still has a couple of hard-coded things, but maybe this is necessary for this more high-level function, we can discuss.

An example of plot produced with the function is attached:
2d_hist_with_proj

### CODE REVIEW

File: plothist/plotters.py [hunk 1]
```diff
@@ -291,18 +291,121 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+    ax_y_projection : matplotlib.axes.Axes
+        The axes for the y projection.
+    ax_colorbar : matplotlib.axes.Axes
+        The axes for the colorbar.
+    """
+
+    colorbar_kwargs.setdefault("label", colorbar_label)
+
+    fig, axes = plt.subplots(
+        figsize=(5.25, 5),
```
Comment (0ctagon) [inline plothist/plotters.py:357]: Why this figsize? I find the projection plots to be a bit too squeezed. Could we increase their height a bit?

File: plothist/plotters.py [hunk 2]
```diff
@@ -291,18 +291,121 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+        pcolormesh_kwargs=pcolormesh_kwargs,
+        colorbar_kwargs=colorbar_kwargs,
+    )
+    plot_hist(hist[:, :: bh.sum], ax=ax_x_projection)
+    plot_hist(hist[:: bh.sum, :], ax=ax_y_projection, orientation="horizontal")
+
+    _ = ax_x_projection.xaxis.set_ticklabels([])
+    _ = ax_y_projection.yaxis.set_ticklabels([])
+
+    ax_y_projection.xaxis.set_ticks_position("top")
```
Comment (0ctagon) [inline plothist/plotters.py:383]: Why this choice? I would prefer the ticks and label to be on the bottom, aligned with the xlabel of the main plot.

File: plothist/plotters.py [hunk 3]
```diff
@@ -291,18 +291,121 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+    ax_y_projection.xaxis.set_label_position("top")
```
Comment (0ctagon) [inline plothist/plotters.py:384]: Same remark, would be better at the bottom of the plot.

File: plothist/plotters.py [hunk 4]
```diff
@@ -291,18 +291,121 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+
+    Parameters
+    ----------
+    hist : 2D boost_histogram.Histogram
+        The 2D histogram to plot.
+    xlabel : str, optional
+        Label for the x axis. Default is None.
+    ylabel : str, optional
+        Label for the y axis. Default is None.
+    ylabel_x_projection : str, optional
```
Comment (0ctagon) [inline plothist/plotters.py:329]: Could we add also a ylabel_y_projection?

**Review by 0ctagon — CHANGES_REQUESTED** [review]

File: plothist/plotters.py [hunk 5]
```diff
@@ -291,18 +291,121 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+    ax_2d.set_ylim(ylim)
+    ax_y_projection.set_ylim(ylim)
+
+    ax_2d.set_xlabel(xlabel)
+    ax_2d.set_ylabel(ylabel)
+    ax_x_projection.set_ylabel(ylabel_x_projection)
+
+    hspace = 0.15
+    wspace = 0.15
+    fig.subplots_adjust(hspace=hspace, wspace=wspace)
```
Comment (0ctagon) [inline plothist/plotters.py:401]: The space between the (`ax_2d` and `ax_x_projection`) and (`ax_2d` and `ax_y_projection`) is not the same.

**Review by 0ctagon — CHANGES_REQUESTED** [review]

File: plothist/plotters.py [hunk 6]
```diff
@@ -353,11 +356,11 @@ def plot_2d_hist_with_projections(
 
     colorbar_kwargs.setdefault("label", colorbar_label)
 
-    fig, axes = plt.subplots(
-        figsize=(5.25, 5),
+    fig, axes = plt.(
```
Comment (0ctagon) [inline plothist/plotters.py:359]: `plt.subplots`

File: docs/basics/2d_hist.rst [hunk 7]
```diff
@@ -155,4 +134,28 @@ Instead of inputing a number of bins and a range in ``make_2d_hist()``, the bins
...
+    fig, ax_2d, ax_x_projection, ax_y_projection, ax_colorbar = plot_2d_hist_with_projections(
+        h,
+        xlabel="variable_0",
+        ylabel="variable_1",
+        ylabel_x_projection="Entries",
+        colorbar_kwargs={"label": "Entries"},
+        save_as="2d_hist_with_projections.svg",
+    )
+
+    fig.savefig("2d_hist_with_projections.svg", bbox_inches='tight')
```
Comment (0ctagon) [inline docs/basics/2d_hist.rst:157]: Either remove this line as we use `save_as` in the function, or keep this line and remove the `save_as`

File: docs/basics/2d_hist.rst [hunk 8]
```diff
@@ -155,4 +134,28 @@ Instead of inputing a number of bins and a range in ``make_2d_hist()``, the bins
...
+To plot a 2d histogram with x and y projections:
+
+.. code-block:: python
+
+    from plothist import make_2d_hist, plot_2d_hist_with_projections
+
+    name_x = "variable_0"
+    name_y = "variable_1"
+
+    h = make_2d_hist([df[df[name_x]%2==0][name_x], df[df[name_x]%2==0][name_y]])
```
Comment (0ctagon) [inline docs/basics/2d_hist.rst:146]: The plot produced by this example is broken. Why this weird definition? 
Suggest to use `h = make_2d_hist([df[name_x], df[name_y]])`

**Review by 0ctagon — CHANGES_REQUESTED** [review]

**Comment by cyrraz** [comment]
```
    # Align ylabels
    # For some reason, fig.align_ylabels() does not work here
    fig.canvas.draw()
    fig.align_ylabels(axs=[ax_2d, ax_x_projection])
    ylabel_position = min(
        ax_2d.yaxis.label.get_position()[0],
        ax_x_projection.yaxis.label.get_position()[0],
    )
    ax_2d.yaxis._autolabelpos = False
    ax_x_projection.yaxis._autolabelpos = False
    ax_2d.yaxis.label.set_position((ylabel_position, 0.5))
    ax_x_projection.yaxis.label.set_position((ylabel_position, 0.5))
```

**Comment by cyrraz** [comment]
Two ways of producing the axes:
1)
```
        fig, axes = plt.subplots(
            figsize=(fig_width, fig_height),
            nrows=2,
            ncols=3,
            gridspec_kw={"width_ratios": [4, 2, 0.25], "height_ratios": [2, 4]},
        )

        ax_2d = axes[1][0]
        ax_x_projection = axes[0][0]
        ax_y_projection = axes[1][1]
        ax_colorbar = axes[1][2]
        axes[0, 1].axis("off")
        axes[0, 2].axis("off")
        hspace = 0.15
        wspace = 0.15
        fig.subplots_adjust(hspace=hspace, wspace=wspace)
```
2)
```
        fig = plt.figure(figsize=(fig_width, fig_height))
        fig_ratio = fig_height / fig_width
        intra_space = 0.05
        margin = 0.15
        height_2d = 0.5
        ax_2d = fig.add_axes(
            [margin * fig_ratio, margin, height_2d * fig_ratio, height_2d]
        )
        ax_x_projection = fig.add_axes(
            [
                margin * fig_ratio,
                margin + height_2d + intra_space,
                height_2d * fig_ratio,
                1 - (2 * margin + height_2d + intra_space),
            ]
        )
        ax_y_projection = fig.add_axes(
            [
                (margin + height_2d + intra_space) * fig_ratio,
                margin,
                (1 - (2 * margin + height_2d + intra_space)) * fig_ratio,
                height_2d,
            ]
        )
        ax_colorbar = fig.add_axes(
            [
                (1 - margin + intra_space) * fig_ratio,
                margin,
                (1 - margin) - (1 - margin + intra_space) * fig_ratio,
                height_2d,
            ]
        )
```

File: plothist/plotters.py (hunk 2 above)
Comment (cyrraz) [reply plothist/plotters.py]: I added a comment in the documentation to explain this choice

File: plothist/plotters.py (hunk 3 above)
Comment (cyrraz) [reply plothist/plotters.py]: Will comment during the meeting

File: plothist/plotters.py (hunk 5 above)
Comment (cyrraz) [reply plothist/plotters.py]: Took 3 hours to fix, but I found a way. The cost of this is that the y labels cannot be aligned.

**Comment by cyrraz** [comment]
New version:

2d_hist_with_projections

File: plothist/plotters.py [hunk 9]
```diff
@@ -291,18 +291,153 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+    pcolormesh_kwargs.setdefault("edgecolors", "face")
     im = ax.pcolormesh(*hist.axes.edges.T, hist.values().T, **pcolormesh_kwargs)
-    divider = make_axes_locatable(ax)
-    cax = divider.append_axes("right", size="5%", pad=0.05)
-    ax.tick_params(axis="x", which="both", top=False, bottom=False)
-    ax.get_figure().colorbar(im, cax=cax, **colorbar_kwargs)
+    ax.get_figure().colorbar(im, cax=ax_colorbar, **colorbar_kwargs)
+
+
+def plot_2d_hist_with_projections(
```
Comment (0ctagon) [inline plothist/plotters.py:309]: Add `offset_xlabel` + documentation

File: plothist/plotters.py [hunk 10]
```diff
@@ -291,18 +291,153 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+
+    fig_width = 6.5
+    fig_height = 6
+
+    fig = plt.figure(figsize=(fig_width, fig_height))
+    fig_ratio = fig_height / fig_width
+    intra_space = 0.05
+    margin = 0.1
+    height_2d = 0.5
+    ax_2d = fig.add_axes([margin * fig_ratio, margin, height_2d * fig_ratio, height_2d])
```
Comment (0ctagon) [inline plothist/plotters.py:376]: Return to old method:
```
fig, axes = plt.subplots(
        figsize=(6.25, 6),
        nrows=2,
        ncols=2,
        gridspec_kw={"width_ratios": [4, 2], "height_ratios": [2, 4]},
    )

    ax_2d = axes[1][0]
    ax_x_projection = axes[0][0]
    ax_y_projection = axes[1][1]
    ax_colorbar = axes[0][1]
    axes[0, 1].axis("off")

    height = (ax_colorbar.get_position().y1 - ax_colorbar.get_position().y0)
    ax_colorbar = fig.add_axes([ax_colorbar.get_position().x0, ax_colorbar.get_position().y0, 0.03, height])
```

Comment (0ctagon) [reply plothist/plotters.py:376]: This also adds the colobar in the top right corner.

File: plothist/plotters.py [hunk 11]
```diff
@@ -291,18 +291,153 @@ def plot_2d_hist(hist, ax, pcolormesh_kwargs={}, colorbar_kwargs={}):
...
+
+
+def plot_2d_hist_with_projections(
+    hist,
+    xlabel=None,
+    ylabel=None,
+    ylabel_x_projection=None,
+    xlabel_y_projection=None,
+    colorbar_label=None,
+    ticks_position_y_projection="top",
```
Comment (0ctagon) [inline plothist/plotters.py:316]: Always bottom, remove argument.

File: plothist/plotters.py (hunk 10 above)
Comment (0ctagon) [reply plothist/plotters.py:376]: Add:

```
    hspace = 0.18
    wspace = 0.18
    fig.subplots_adjust(hspace=hspace, wspace=wspace)

    fig.align_ylabels()

    if save_as is not None:
        fig.savefig(save_as, bbox_inches="tight")
 ```

**Review by 0ctagon — CHANGES_REQUESTED** [review]

File: docs/basics/2d_hist.rst [hunk 12]
```diff
@@ -143,7 +143,7 @@ To plot a 2d histogram with x and y projections:
     name_x = "variable_0"
     name_y = "variable_1"
 
-    h = make_2d_hist([df[df[name_x]%2==0][name_x], df[df[name_x]%2==0][name_y]])
+    h = make_2d_hist([df[name_x], df[name_y]])
```
Comment (0ctagon) [inline docs/basics/2d_hist.rst:146]: Add more bins (at least 50)

File: docs/basics/2d_hist.rst [hunk 13]
```diff
@@ -151,7 +151,6 @@ To plot a 2d histogram with x and y projections:
         ylabel="variable_1",
         ylabel_x_projection="Entries",
         colorbar_kwargs={"label": "Entries"},
-        save_as="2d_hist_with_projections.svg",
     )
 
     fig.savefig("2d_hist_with_projections.svg", bbox_inches='tight')
```
Comment (0ctagon) [inline docs/basics/2d_hist.rst:156]: Keep `save_as`

**Comment by 0ctagon** [comment]
In the `README.rst`, l.33 remove `https://raw.githubusercontent.com/cyrraz/plothist/main/docs/img/intro_2dhist_complexe.svg` and add `https://raw.githubusercontent.com/cyrraz/plothist/main/docs/img/NAME_OF_THE_NEW_PLOT.svg`. This last link will work after merging.

**Comment by 0ctagon** [comment]
Also also `v0.7.2` -> `v0.8`

**Comment by cyrraz** [comment]
New version:

2d_hist_with_projections

**Review by 0ctagon — APPROVED** [review]
Beautiful.

**Pull request merged**